### PR TITLE
SEAB-5913: Improve search relevance

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR improves the relevance of Dockstore search results.

To run a search, the UI synthesizes a search query using the bodybuilder library and submits it to ES.  Before continuing, please read the following, which explains the structure of typical ES search queries:

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html

In the previous implementation:
* Keyword subqueries were run in a `filter` context, preventing ES from scoring the results.
* The UI always sorted the results, per the header control settings, so even if ES returned results sorted by relevance, the UI would reorder them (in descending star order by default).
* Essentially, when we ran a search, ES was determining all of the hits that matched the search criteria, sorting them by stars, and returning an arbitrary 200 of them.  There was no ordering by relevance, so for searches with many hits, it was entirely possible that the most relevant entry could be omitted from the results.

In this implementation:
* Keyword subqueries run in a `should` context, which allows ES to calculate a relevance score for each hit by combining the scores of the subqueries and order the returned hits by score.  Each hit's score is returned in the query response.
* We change the UI so sorting is initially disabled, and the results are displayed in the order that ES returns them.  Each header sort control now has a "none" setting, in addition to "ascending" and "descending".
* When we run a keyword search,  ES now orders the hits by score.  Otherwise, it orders the results by stars as it did before.
* The query synthesizer now splits the search string into terms, and creates individual subqueries for each term.   The net effect is that entries which contain both terms are marked as more relevant than before.
* To improve relevance, weights have been introduced, via the `boost` setting in each subquery.
* The topic and categories are now included in keyword subqueries.

When reviewing the code, keep in mind that bodybuilder's `orQuery` adds a subquery to the `should` context, and that we force a minimum number of "shoulds" to match via the `minimum_should_match` field.

There is a companion PR that tweaks the webservice to further improve search relevance:
https://github.com/dockstore/dockstore/pull/5686

In combination, these two PRs greatly improve the quality of Dockstore keyword search results.  On prod, Ash's "samtools index" keyword search test ranks his and Milan's workflows, which are probably the most relevant hits, somewhere between the middle and bottom of all results.

After the improvements, his and Milan's workflows appear as results 1 and 3, with Julian's, which is also quite relevant, as result 2:

image

Yay!

**Review Instructions**
Go to Search, and run some keyword, facet, and advanced searches, looking for inconsistencies in the behavior or results, or regressions in search quality.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-5913

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

### CODE REVIEW

File: src/app/search/query-builder.service.ts [hunk 1]
```diff
@@ -221,16 +247,23 @@ export class QueryBuilderService {
...
-        .orFilter('match_phrase', 'tags.sourceFiles.content', searchString)
-        .orFilter('match_phrase', 'description', searchString)
-        .orFilter('match_phrase', 'labels', searchString)
-        .orFilter('match_phrase', 'author', searchString)
-    );
+    // Extract each search term from the search string, limiting to a maximum of 20 terms to prevent a DOS attack
+    const terms = searchString.trim().split(' ').slice(0, 20);
+    terms.forEach((term) => {
+      body
+        .orQuery('wildcard', 'full_workflow_path', { value: '*' + term + '*', case_insensitive: true, boost: 7 })
```
Comment (denis-yuen) [inline src/app/search/query-builder.service.ts:254]: Neat! The boost categories are pretty readable/tweakable in the future

**Review by denis-yuen — APPROVED** [review]
:partying_face:

**Review by kathy-t — APPROVED** [review]

File: src/app/search/query-builder.service.ts [hunk 2]
```diff
@@ -191,23 +195,45 @@ export class QueryBuilderService {
...
+        body = this.searchEverything(body, values);
+      } else {
+        body = body.query('match_all', {});
+      }
     }
     return body;
   }
 
+  private isEmpty(values: string): boolean {
+    return values == undefined || values.toString().trim().length <= 0;
```
Comment (coverbeck) [inline src/app/search/query-builder.service.ts:217]: Nit: `values === undefined`, unless you're trying to catch the null case as well, in which case `!values` is more readable to me.

File: src/app/search/query-builder.service.ts [hunk 3]
```diff
@@ -261,90 +294,38 @@ export class QueryBuilderService {
     return body;
   }
 
-  // Advanced search query related functions
+  advancedSearchDescription(body: any, advancedSearchObject: AdvancedSearchObject) {
+    this.advancedSearch(body, advancedSearchObject, 'description');
+  }
 
-  /* TODO: Make this better */
```
Comment (coverbeck) [inline src/app/search/query-builder.service.ts:266]: You made it better! Less code is good.

**Review by coverbeck — APPROVED** [review]
> Each header sort control now has a "none" setting, in addition to "ascending" and "descending"

Just curious, where is this in the code? Is it because you don't explicitly set the `start` attribute?

This is great; it seems like we've only been using of a small subset of the ES feature set until now.

**Review by david4096 — APPROVED** [review]

**Comment by svonworl** [comment]
> > Each header sort control now has a "none" setting, in addition to "ascending" and "descending"
> 
> Just curious, where is this in the code? Is it because you don't explicitly set the `start` attribute?

Previously, we disabled the "none" setting with `matSortDisableClear`:
https://material.angular.io/components/sort/overview

**Pull request merged**